x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/TypedChainTaskExecutor.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.utils;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.util.concurrent.AbstractRunnable;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Predicate;
+
+/**
+ * A utility that allows chained (serial) execution of a number of tasks
+ * in async manner.
+ */
+public class TypedChainTaskExecutor<T> {
```

Comment:
I wonder if we can now remove `ChainTaskExecutor` and reuse this one. It would definitely be nice to only need one of them if it's possible. I haven't looked how hard it'd be, feel free to ping me to discuss.

Reply:
Definitely, originally it was an abstract class that ChainTaskExecutor inherited from and simply supplied defaults. I could look a little bit into cleaning up the two classes.

Reply:
@dimitris-athanasiou I have made the typed chain task executor the base class as it provides additional options, and renamed `ChainTaskExecutor` to `VoidChainTaskExecutor` and the vast majority of all the interfaces with that Class can stay the same. Some minor switches with return times as I do collect all the responses in a `List`